internal/engine/netx/quicdialer/errorwrapper_test.go

```diff
@@ -43,6 +43,28 @@ func errorWrapperCheckErr(t *testing.T, err error, op string) {
 		t.Fatal("unexpected failure")
 	}
 }
+func TestErrorWrapperInvalidCertificate(t *testing.T) {
+	nextprotos := []string{"h3"}
+	servername := "example.com"
+	tlsConf := &tls.Config{
+		NextProtos: nextprotos,
+		ServerName: servername,
+	}
+
+	dlr := quicdialer.ErrorWrapperDialer{Dialer: &quicdialer.SystemDialer{}}
+	// use Google IP
+	sess, err := dlr.DialContext(context.Background(), "udp",
+		"216.58.212.164:443", tlsConf, &quic.Config{})
+	if err == nil {
+		t.Fatal("expected an error here")
+	}
+	if sess != nil {
+		t.Fatal("expected nil sess here")
+	}
+	if err.Error() != errorx.FailureSSLInvalidCertificate {
+		t.Fatal("unexpected failure")
+	}
+}
```

Comment:
Side note: the fact that we need to write tests for this functionality we're adding in `errorx` in _another_ package does seem to indicate that perhaps part of the classifier should live here instead. (There will _always_ be some cross dependency since `errorx` contains the errno classifier but it seems to me moving the QUIC specific classifier here would reduce the coupling, and hence enhance maintainability.)

Would you mind adding a comment in `errorx.go` mentioning that perhaps we want to do that in the future?

Reply:
I'm not sure if I agree that it will be better to move them. But I added the comment so we can keep that in mind :)

Reply:
Thanks! Why do you disagree with moving them?

Reply:
Well, these classifiers are used to translate a given error to a string constant which is defined in `errorx`. So I think that if `errorx` introduces and defines those OONI error strings, it makes sense that `errorx` also handles the translation/mapping. It seems wrong (at least to me) to let the transport layer packages handle this mapping... :thinking:

Reply:
I do see the point that it does not seem very elegant to use the `errorx` string as a check in `quicdialer/errorwrapper_test`. Saying that, I realize, that I aimed to do an integration test of a "real" QUIC failure, as you mentioned above, with this test.

Reply:
I see what you mean, and that's a good point I also see myself!

The flip side is that you need to write tests in, say, `quicdialer` to ensure that the mapping that is defined in `errorx` is working correctly for real networking scenarios. At the same time, when you want to, say, add a new error for QUIC, you need to modify code in `quicdialer` and then `errorx` and then make sure that they are in sync, which kinda breaks the nice property that packages should be as self-contained as possible.

It may be that it's more convenient to have everything quic-errors related in `quicdialer` and to use `errorx` only as a support package to classify the errors that we don't otherwise know how to classify. Even this solution, though, may not be ideal considering that we would still probably have some functionality here and some functionality there.

Given it's not clear what's best, the best course of action is just documenting this as something we may wanna consider for the future, next time we will need to apply changes to this part of the tree.

Reply:
yes :+1: I understand. It's difficult to clearly separate the functionalities here.